src/core/reference/src/op/paged_attention.cpp

```diff
@@ -0,0 +1,282 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "openvino/reference/paged_attention.hpp"
+
+#include <cmath>
+#include <cstdint>
+#include <cstring>
+#include <ctime>
+#include <memory>
+#include <random>
+#include <stdexcept>
+#include <vector>
+
+#include "openvino/core/shape.hpp"
+
+// ==== HELPER FUNCS ====
+template <typename T>
+T dot_product(const T* a, const T* b, int32_t size) {
+    T sum = T(0);
+    for (int32_t i = 0; i < size; i++)
+        sum += a[i] * b[i];
+    return sum;
+}
+
+template <typename T>
+void softmax(std::vector<T>& scores) {
+    T max_score = *std::max_element(scores.begin(), scores.end());
+    T sum = T(0);
+    for (auto& s : scores) {
+        s = std::exp(s - max_score);
+        sum += s;
+    }
+    for (auto& s : scores)
+        s /= sum;
+}
+
+/*
+ * Apply RoPE (Rotary Positional Embedding) rotation to a vector.
+ */
+template <typename T>
+void apply_rope(T* vec, int32_t head_size, const T* rotation_trig_lut, int32_t trig_index) {
+    int32_t half = head_size / 2;
+    // Each row is of length head_size.
+    const float* row = rotation_trig_lut + trig_index * head_size;
+    for (int32_t i = 0; i < half; i++) {
+        T x0 = vec[2 * i];
+        T x1 = vec[2 * i + 1];
+        T cos_val = row[i];
+        T sin_val = row[half + i];
+        vec[2 * i] = x0 * cos_val - x1 * sin_val;
+        vec[2 * i + 1] = x0 * sin_val + x1 * cos_val;
+    }
+}
+
+/*
+ * Helper to check whether a given block_id is present in rotated_block_indices.
+ * If found, returns true and sets rotated_index to the position within rotated_block_indices.
+ */
+bool get_rotated_index(int32_t block_id,
+                       const int32_t* rotated_block_indices,
+                       int32_t num_rotated_blocks,
+                       int32_t& rotated_index) {
+    for (int32_t i = 0; i < num_rotated_blocks; i++) {
+        if (rotated_block_indices[i] == block_id) {
+            rotated_index = i;
+            return true;
+        }
+    }
+    return false;
+}
+
+namespace ov {
+namespace reference {
+
+template <typename T>
+void paged_attention(
```

Comment:
I suppose that this function will be further divided into functions or objects and be covered with unit tests? The functional testing I see at this stage is quite minimal already (somewhat understandable, since there is a lot of parameters to cover), but  proper *unit* testing in reference-grade code is a must anyway, especially in view of such sparse functional testing.

Reply:
I have added back previous tests with newly calculated RoPE. I did try to separate the function into pieces a bit more, albeit I have a feeling doing it more will make the code more confusing at this point - I can try to add these in a separate PR